core/storage/feedback/gae_models.py

```diff
@@ -420,6 +421,400 @@ def get_multi(cls, user_id, thread_ids):
+    # on running the job for calculating the number of messages in a thread
+    # and updating the message_count field we do not wish the last_updated field
+    # to be updated.
+    last_updated = ndb.DateTimeProperty(indexed=True)
+
+    def put(self, update_last_updated_time=True):
+        """Writes the given thread instance to the datastore.
+
+        Args:
+            update_last_updated_time: bool. Whether to update the
+                last_updated_field of the thread.
+
+        Returns:
+            thread. The thread entity.
+        """
+        if update_last_updated_time:
+            self.last_updated = datetime.datetime.utcnow()
+
+        return super(GeneralFeedbackThreadModel, self).put()
+
+
+    @classmethod
+    def generate_new_thread_id(cls, entity_type, entity_id):
+        """Generates a new thread ID which is unique.
+
+        Args:
+            entity_type: str. The type of the entity.
+            entity_id: str. The ID of the entity.
+
+        Returns:
+            str. A thread ID that is different from the IDs of all
+                the existing threads within the given entity.
+
+        Raises:
+           Exception: There were too many collisions with existing thread IDs
+               when attempting to generate a new thread ID.
+        """
+        for _ in range(_MAX_RETRIES):
+            thread_id = (
+                entity_type + '.' + entity_id + '.' +
+                utils.base64_from_int(utils.get_current_time_in_millisecs()) +
+                utils.base64_from_int(utils.get_random_int(_RAND_RANGE)))
+            if not cls.get_by_id(thread_id):
+                return thread_id
+        raise Exception(
+            'New thread id generator is producing too many collisions.')
+
+    @classmethod
+    def create(cls, thread_id):
+        """Creates a new FeedbackThreadModel entry.
+
+        Args:
+            thread_id: str. Thread ID of the newly-created thread.
+
+        Returns:
+            GeneralFeedbackThreadModel. The newly created FeedbackThreadModel
+                instance.
+
+        Raises:
+            Exception: A thread with the given thread ID exists already.
+        """
+        if cls.get_by_id(thread_id):
+            raise Exception('Feedback thread ID conflict on create.')
+        return cls(id=thread_id)
+
+    @classmethod
+    def get_threads(
+            cls, entity_type, entity_id, limit=feconf.DEFAULT_QUERY_LIMIT):
+        """Returns a list of threads associated with the entity, ordered
+        by their "last updated" field. The number of entities fetched is
+        limited by the `limit` argument to this method, whose default
+        value is equal to the default query limit.
+
+        Args:
+            entity_type: str. The type of the entity.
+            entity_id: str. The ID of the entity.
+            limit: int. The maximum possible number of items
+                in the returned list.
```

Comment:
nit: some words can be moved to the previous line.

Reply:
Done